Importer UI overhaul rebase/update

davidswarbrick opened the pull request:
## Description

Continuing work on PR #1685 implementing an improved UI, with more discussion/screenshots at #1593.
Previous work has been rebased on `beetbox/beets/master`. Currently I've fixed many of the tests which were failing due to the code targeting Python 2.7 (such as `isinstance(basestring)`) taking care to retain backwards compatibility so that we can get this into release 1.5.0. The current tests that are failing are asserting the output of the UI, ~I'm not sure if that's because they are yet to be updated for the new UI but that shouldn't be hard to address if so~ **Update:** The failing tests were due to a change in UI. I've adjusted the failing tests to match the new UI output, however there are extra tests that must be written to check, for example, Unicode output of the "not equals" sign if we wish to adopt the new UI.

Once all tests are passing a substantial reformat is needed to adhere to `flake8` rules, I'd prefer to do this with black in one commit but that might be best to do after #3694 is merged.

## To Do

- [x] Rebased on current `master` branch
- [x] Fix errors due to Python 2.7/3 discrepancies / internal logic
- [x] Investigate whether failing tests need to be updated for new UI
- [x] Review code for efficiency/legibility
- [x] Add any extra tests for functionality not currently covered (e.g. construction of columns, Unicode output)
- [x] Document functions & add whitespace etc. for compliance with `flake8` (might be easiest to automate whitespace with black)
- [x] Document differences to previous UI, perhaps with screenshots etc.
- [x] Changelog - Add an entry to `docs/changelog.rst` near the top of the document

jtpavlock commented:
Wow! This would be awesome to finally get finished! Thanks for taking this on. 

Instead or targeting the ui branch, I think if you target master it will help you find actual conflicts and make the diff much more manageable

RollingStar commented:
This is amazing.

sampsyo commented:
I would love to keep this overhaul in play.

sampsyo commented:
Yes, let's keep it open.

davidswarbrick commented:
Tests passing because the ui tests haven't been rewritten yet!

davidswarbrick commented:
After some hefty re-working of the great work by @mxmerz (now over 5 years ago!) I'm hopeful that these changes are ready for a code review, and some documentation / publicity. If anyone would like to help add some screenshots from their setups it could be a good opportunity for non-code contributions to beets!

I've moved the vast majority of the new UI into the `ChangeRepresentation` class @mxmerz started, including making "newline" and "column" layouts into generic functions, which improves legibility and re-usability. This has allowed implementing column/newline layouts at the album level, not just for tracklists! We may wish to move those into `ui/__init__.py` in the future. I've added tests to check the column and newline wrapping works on each version of python, and documentation to the functions to explain their data structures and differences in output. As a consequence of "genericising" the layout printing functions, I've removed some of the custom data structures implemented originally which should help reviewers and future contributors alike understand what each function does.

The changes are best understood via `show_change()` and `show_item_change()` in `beets/ui/commands.py` which trigger the functions to print changes to the CLI. The docstrings for `print_column_layout` and `print_newline_layout` explain how the layouts are generated from the provided dictionaries. In the case of tracklists (which previously implemented a custom formatting structure) the function calls are now simpler: `show_match_tracks` -> `make_line` (for each track) -> `print_tracklist` (for all tracks) .
As always, happy for any comments or suggestions. Hope we can finally get this moving into main!

sampsyo commented:
Absolutely incredible work! Thank you for reviving this effort. I would love to merge this in when we're confident in it!

Would you mind doing a little publicity to encourage some extra pairs of eyes on the new UI? As you mentioned, getting people to try it out and take screenshots would be an awesome non-code contribution to the project. Perhaps you could make a post or two in the GitHub Discussions area and/or on Discourse to help encourage people to take a look?

seitzbg commented:
This is amazing, any ETA for a merge?

davidswarbrick commented:
Hi, I'd hope this is still relevant, although this PR has indeed gone stale. 
I've been using this on a personal branch for over six months now, but it would be good to get some other community involvement in the form of other testers or a review.

davidswarbrick commented:
Rebased on latest master, and removed the unnecessary lines from the diff (blank lines or apostrophe substitutions). Happy for this to be reviewed and merged!

wisp3rwind commented:
> Rebased on latest master, and removed the unnecessary lines from the diff (blank lines or apostrophe substitutions). Happy for this to be reviewed and merged!

I have to confess that I still haven't tested this myself; but I'll try to gradually read the code, hopefully sooner than later.

One small thing upfront: I think we can drop all the explicit unicode prefixes `u"..."` for string literals, since Python 2 support was dropped in the meantime.

wisp3rwind commented on beets/ui/__init__.py:
```diff
@@ -465,91 +478,192 @@ def human_seconds_short(interval):
     """
-    if not isinstance(a, str) \
-       or not isinstance(b, str):
+    # Set highlight colors.
+    highlight_added = "text_diff_added"
+    highlight_removed = "text_diff_removed"
+    minor_highlight = "text_highlight_minor"
```
Wouldn't it be simpler to use the abstract color names directly below instead of having these variables? The color names are just as descriptive and easy to replace/search for/etc if needed.

davidswarbrick replied:
Yes, have done so.

wisp3rwind commented on beets/ui/__init__.py:
```diff
@@ -465,91 +478,192 @@ def human_seconds_short(interval):
+    minor_highlight = "text_highlight_minor"
+
+    if not isinstance(a, str) or not isinstance(b, str):
         # Non-strings: use ordinary equality.
```
This is what the code did before, but code and comment are in conflict (since the stringified objects are compared). Maybe one should check the `git blame` to figure out what the original intention was here? I would guess that the right thing to do is what the comment says, but still return the stringified (and possibly highlighted) versions.

davidswarbrick replied:
Git blame did indeed show that two different people added the comment and functionality. The original intention seems to be that e.g. ints are compared as-is. I've moved the bytes -> readable paths upwards, and then kept the catch-all "not strings" comparison to recreate the original functionality.

wisp3rwind reviewed (changes requested):
Added a first round of comments, see the inline notes.

wisp3rwind commented on beets/ui/__init__.py:
```diff
@@ -559,6 +673,16 @@ def _colordiff(a, b, highlight='text_highlight',
     b_out = []
 
+    add_mapper = (
+        lambda w: w if re.match(r"(\s)", w) else colorize(highlight_added, w)
+    )
+    remove_mapper = (
+        lambda w: w if re.match(r"(\s)", w) else colorize(highlight_removed, w)
+    )
+    minor_mapper = (
+        lambda w: w if re.match(r"(\s)", w) else colorize(minor_highlight, w)
+    )
```
These could use [`w.isspace()`](https://docs.python.org/3.10/library/stdtypes.html#str.isspace) instead of resorting to full-blown regexes.

<del>Additionally, wouldn't it make sense to move the check for empty or whitespace-only strings into `colorize` itself, and get rid of the `*_mapper`s completely?</del> Nevermind, I misunderstood the purpose of having these mappers (only) in this function. However, see below for another code suggestion.

davidswarbrick replied:
I'm re-reviewing this myself (it was inherited from the previous maintainer of this PR) and I agree that these regexes are confusing and/or unnecessary. At one point I think I understood why these mappers existed but now I've arrived at your conclusion - why can't we check for empty whitespace elsewhere? I notice you have an answer but would appreciate it if you could share! 

In the meantime I did try your suggestion below and unfortunately re.Match objects aren't iterable, although the .isspace() option will simplify this somewhat anyway.

wisp3rwind replied:
> At one point I think I understood why these mappers existed but now I've arrived at your conclusion - why can't we check for empty whitespace elsewhere? I notice you have an answer but would appreciate it if you could share!

I believe the point is that it's not always clear whether we want to "highlight" whitespace (note that the background color might be different depending on whether colorization is applied or not. That's generally a design decision that may depend on context. Here, whoever implemented this first chose to not highlight whitespace (changes). But not every caller of `colorize` might want that behaviour.

> In the meantime I did try your suggestion below and unfortunately re.Match objects aren't iterable, although the .isspace() option will simplify this somewhat anyway.

What exactly did you try? There's a small error in my suggestion below, it should have been

```python
b_out.append(re.sub(r"\S+", lambda s: colorize(s, highlight_added), b))
```
note the added argument `b`.

Alternatively, I just thought of a different way to achive the conditional whitespace highlighting: Add a keyword argument to `colorize`, as in `def colorize(color_name, text, ws_highlight=True)` and handle the `ws_highlight=False` case in `colorize`.

wisp3rwind replied:
I've had another look at the `re` suggestion, and I see what you meant when saying that `re.Match` objects aren't iterable. It's easy to solve though:
```
b_out.append(re.sub(r"\S+", lambda s: colorize(s.group(0), highlight_added), b))
```
cf. the example here: https://docs.python.org/3.10/library/stdtypes.html#str.title

davidswarbrick replied:
I've simplified this by implementing your suggestion of moving whitespace highlighting to colorize :)

wisp3rwind commented on beets/ui/__init__.py:
```diff
@@ -567,31 +691,39 @@ def _colordiff(a, b, highlight='text_highlight',
         elif op == 'insert':
             # Right only.
-            b_out.append(colorize(highlight, b[b_start:b_end]))
+            words = re.split(r"(\s)", b[b_start:b_end])
+            words_colorized = map(add_mapper, words)
+            b_out.append("".join(words_colorized))
         elif op == 'delete':
```
I think (didn't test), these constructions could be replaced by the following (simpler, in my opinion) code, which would also remove the need for the somewhat opaque `*_mapper`s:

```python
b_out.append(re.sub(r"\S+", lambda s: colorize(s, highlight_added)))
```

which should replace all matches of the regex (non-space characters in this case) by the result of applying the function in the second argument, cf. https://docs.python.org/3/howto/regex.html#search-and-replace.

wisp3rwind commented on beets/ui/__init__.py:
```diff
@@ -648,6 +780,269 @@ def term_width():
     return width
 
 
+def split_into_lines(string, width_tuple):
```
Not a mandatory thing, but this looks like a function which it would be nice to have a unittest for, which checks a few of the edge case that the previous version of the code (as mentioned in the code comment below) failed to handle properly.

davidswarbrick replied:
This turned out to be important! By implementing a unit test I found errors in the colored text handling, which I have now fixed.

wisp3rwind reviewed (changes requested):
... and a few more suggestions :)

davidswarbrick commented on beets/ui/__init__.py:
```diff
@@ -465,91 +478,192 @@ def human_seconds_short(interval):
+        if code in ANSI_CODES.keys():
+            escape = escape + COLOR_ESCAPE + "%im" % ANSI_CODES[code]
+        else:
+            raise ValueError("no such ANSI code %s", code)
```
Agree, I've moved this to the config checking now.

davidswarbrick commented on beets/ui/__init__.py:
```diff
@@ -465,91 +478,192 @@ def human_seconds_short(interval):
+                    else:
+                        raise ValueError("no such color %s", color_def)
+
+                COLORS[name] = color_def
```
This is a great idea, I've moved this code around.

wisp3rwind reviewed (commented):
Two small nitpicks on the new changes, otherwise they're looking great!

davidswarbrick commented on beets/ui/__init__.py:
```diff
@@ -651,67 +656,53 @@ def _colordiff(a, b):
-        a = util.displayable_path(a)
-        b = util.displayable_path(b)
+            return colorize("text_diff_removed", str(a)),
+            colorize("text_diff_added", str(b))
```
agreed, implemented.

davidswarbrick commented on beets/ui/__init__.py:
```diff
@@ -651,67 +656,53 @@ def _colordiff(a, b):
-        a = str(a)
-        b = str(b)
         if a == b:
             return a, b
```
Agreed, implemented.

davidswarbrick commented:
hey @wisp3rwind I have addressed the comments you made.
I see from #4488 and [the discourse discussion of a roadmap](https://discourse.beets.io/t/project-roadmap/1291/3?u=adrian) that there are lots of great but longer-term ideas around UI/GUI/output rendering for beets, however I see this PR as a separate and completed improvement for short-term usability as we work towards new interface options, what do you think?

wisp3rwind commented:
> hey @wisp3rwind I have addressed the comments you made. I see from #4488 and [the discourse discussion of a roadmap](https://discourse.beets.io/t/project-roadmap/1291/3?u=adrian) that there are lots of great but longer-term ideas around UI/GUI/output rendering for beets, however I see this PR as a separate and completed improvement for short-term usability as we work towards new interface options, what do you think?

Hi, yeah, I didn't forget this PR; I'm just struggling to review it properly — in fact #4488 is not really a sign of big plans ahead, but merely an attempt to build a tool that help with the review of this PR.

I agree, this PR is more of an incremental change, and should be merged rather sooner than later. There seems to be another conflicting PR, #4473, which seems to be more disruptive (not sure; didn't read the code; also stalled for now) and might warrant more discussion on whether we want it.

JOJ0 commented:
@davidswarbrick Hi! I changed this PR back to draft state even though it seems to be very far in the process. Please if you set it back to "ready for review" if you find time to continue/finalize it. Hope that's ok for you and drop us a line any time! :-) Thanks for all your efforts!

davidswarbrick commented:
Hi @JOJ0, it's quite frustrating to see this moved back to a draft PR. It is ready for review and has been for some time as the comments above indicate. My understanding is that the delay in merge is due to incomplete previous reviews. I've responded to the initial comments but have received no further review comments or points to address, so I don't understand what is required to get this merged, or whether there is any desire to do so. Please could you specify what you expect me to add?

JOJ0 commented:
Hi @davidswarbrick I'm very sorry, that was my bad! I didn't read thoroughly enough but saw unchecked boxes and currently am looking to tidy up the loads of PR's and try to find out what is stalling and why. If you think this PR is ready to review please set it back to that state. I really didn't want to add add to frustration your efforts are very much appreciated. The problem currently in the beets project is that most maintainers don't find time to review properly. I'm not sure how I can help at the moment but will think about it and bring up the topic again with maintainers!

davidswarbrick commented:
I appreciate the explainer & apology, thanks @JOJ0, it's a huge project to work on and I'm grateful to all the maintainers who have given their time to look at this! 
I've just squashed and rebased on the latest master, please do let me know if there is anything I can do to help facilitate a code review - for example I'd be happy to get on a call with a maintainer if that would be helpful. 
The only outstanding checkbox is the changelog entry - I can draft something but I expect a UI change might come with a larger announcement. Since this PR has stalled for so long I haven't been sure if there's a reluctance to have such a large code change (as it clearly is difficult to review), which is why I haven't prepared that documentation, but if the maintainers would like to go ahead with this PR I'd be happy to flesh out the documentation and draft a release note.

JOJ0 commented:
Hi @davidswarbrick, I'm finally testing this gui in my daily workflows. I play with configurations right now. I don't get what the difference between `column` and `newline` lineouts is. Can you help me? What exactly would change if I do this for example:

```
ui:
    terminal_width: 40
    color: yes
    colors:
        action: ['bold', 'blue']
        action_description: ['normal']
        # import_path: ['bold', 'inverse', 'white']
        #import_path: ['normal']
    import:
        layout: newline
```

JOJ0 commented:
Ah, reading the docstrings of `print_column_layout` and `print_newline_layout` right now! The difference is documented here. Also I understand that `terminal_width` is a fallback only. To really see a wrapping, I need to adjust my actual terminal width. Ok ok, nevermind, thanks anyway!

JOJ0 commented:
I'm not sure, but I think I found an issue:
![image](https://user-images.githubusercontent.com/2733783/224477845-9ddbff55-9ca3-45f8-a1b7-095240a0bcd0.png)

Seems like sometimes there are redundant whitespace characters thrown into the color-coded-text-diff.

davidswarbrick commented:
Thanks for the review! Yes I agree this looks like a bug, I think I noticed it too the other day but wasn’t sure if it was my terminal config. I’ll investigate further and update when I get the chance.On 11 Mar 2023, at 09:59, J0J0 Todos ***@***.***> wrote:﻿
I'm not sure, but I think I found an issue:

Seems like sometimes there are redundant whitespace characters thrown into the color-coded-text-diff.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

davidswarbrick commented:
Hey @JOJ0, I've implemented some extra test suites which should catch the bug you found, they're passing now after some improvements, so if you get the chance it would be good to see if your testing works :smile:

JOJ0 commented:
Hi @davidswarbrick I have not forgotten about this PR. Thanks for the bugfix. I would like to finally integrate the current state into my personal main branch and try this new UI in my day to day use. 

A couple of things that would help me move on:
- Could you rebase some of your latest commits into fewer commits so it's easier to cherry-pick. But please keep them separated if you think it makes sense and explains how and why you did things! maybe just integrate the linting commits into the original "feature commits" or something like that. tiny cleanups would help already!
- There is some UI code I was hacking on recently - it now conflicts with yours. It shouldn't be too much trouble to integrate it.
- If you don't find the time at all for this cleanup work - do you mind if I hijack your branch and do these things?


And then my further idea would be to get some more feedback:
- Create a beets discussion and call for beets users that want to test and give feedback on the new UI.
- I would create a simple how-to on how to get user your feature branch or even cherry-pick the changes.
- I would just like to get a few more people to report that this is working fine for a few weeks (I actually did that an except the bug you fixed recently all went fine)

davidswarbrick commented:
Thanks @JOJ0 for the suggestions, I've squashed my recent commits into a single one representing the bugfix for the issue you found to help re-reviews, but feel free to hijack and squash further if we are to merge. I've rebased on the latest master, fixing the merge conflict you identified.
I completely agree about "real human testing", but thanks for recognising the work finding edge-cases :smile: I'll draft a discussion post!

davidswarbrick commented:
In response to[ user feedback](https://github.com/beetbox/beets/discussions/4870#discussioncomment-6892885) from @arsaboo I've made a minor change to the default config which will improve readability in PuTTY Shells (as Windows users may use this or a derivative to access their beets instance). Those who wish to use the colour scheme from my previous screenshots can easily override the defaults in their personal `config.yaml`.

arsaboo commented:
Thanks @davidswarbrick for the fix....just tried it and it looks great 💯 

<img width="490" alt="image" src="https://github.com/beetbox/beets/assets/18319734/f0740109-812c-41aa-b2f6-891447f43313">

wazabees commented:
Posting here as I got an error message when testing out beets.git@ui.

My OS is Arch Linux and I use pipx manage python packages. First I uninstalled beets, and the injected plugins.

I then installed beets.git@ui using `pipx install git+https://github.com/davidswarbrick/beets.git@ui`. After installation, I could use beets normally to search my music, seemingly most commands worked as expected, except for import.

Using `beet import` gave me the following error:

```
Traceback (most recent call last):
  File "/home/wazabees/.local/bin/beet", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/wazabees/.local/pipx/venvs/beets/lib/python3.11/site-packages/beets/ui/__init__.py", line 1782, in main
    _raw_main(args)
  File "/home/wazabees/.local/pipx/venvs/beets/lib/python3.11/site-packages/beets/ui/__init__.py", line 1769, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/home/wazabees/.local/pipx/venvs/beets/lib/python3.11/site-packages/beets/ui/commands.py", line 1285, in import_func
    import_files(lib, paths, query)
  File "/home/wazabees/.local/pipx/venvs/beets/lib/python3.11/site-packages/beets/ui/commands.py", line 1225, in import_files
    session.run()
  File "/home/wazabees/.local/pipx/venvs/beets/lib/python3.11/site-packages/beets/importer.py", line 353, in run
    pl.run_parallel(QUEUE_SIZE)
  File "/home/wazabees/.local/pipx/venvs/beets/lib/python3.11/site-packages/beets/util/pipeline.py", line 446, in run_parallel
    raise exc_info[1].with_traceback(exc_info[2])
  File "/home/wazabees/.local/pipx/venvs/beets/lib/python3.11/site-packages/beets/util/pipeline.py", line 311, in run
    out = self.coro.send(msg)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/wazabees/.local/pipx/venvs/beets/lib/python3.11/site-packages/beets/util/pipeline.py", line 170, in coro
    task = func(*(args + (task,)))
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wazabees/.local/pipx/venvs/beets/lib/python3.11/site-packages/beets/importer.py", line 1469, in user_query
    task.choose_match(session)
```

Please let me know if there is anything else I can do to help figure this out. I'm not very experienced with python - or bug-hunting. Thanks! ;)

JOJ0 commented:
thanks @wazabees for the report, that helps!

I think your error text is truncated at the bottom. The ansi something error msg is missing. Might have been my mistake when quoting it in our PM discussion on beets.discourse. Sorry! Please update your post with full error. Many thanks!

wazabees commented:
> thanks @wazabees for the report, that helps!
> 
> I think your error text is truncated at the bottom. The ansi something error msg is missing. Might have been my mistake when quoting it in our PM discussion on beets.discourse. Sorry! Please update your post with full error. Many thanks!

Updated!

RollingStar commented:
I have the same issue as @wazabees. My setup is pretty janky, and it's on ARM, so it could be an edge case. Info about my setup here: https://github.com/beetbox/beets/discussions/4910

```
root@77ac896c76f0:/#beet -vv import /temp/data/use/music/imp/
user configuration: /config/config.yaml
data directory: /config
plugin paths:
Sending event: pluginload
library database: /blk/black/mysys/beet/beets_library.db
library directory: /blk/black/og
Sending event: library_opened
Sending event: import_begin
Sending event: import_task_created
Sending event: import_task_start
Looking up: /temp/data/use/music/imp; /temp/data/use/music/imp/The Beach Boys - 1966 - Pet Sounds (Japan SHM-SACD 2014) [flac]
Tagging The Beach Boys - Pet Sounds
No album ID found.
Search terms: The Beach Boys - Pet Sounds
Album might be VA: False
Searching for MusicBrainz releases with: {'release': 'pet sounds', 'artist': 'the beach boys', 'tracks': '13'}
Requesting MusicBrainz release 2c2c80be-57e3-44d2-a910-a7a873683141
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_album_extract
Sending event: albuminfo_received
Candidate: The Beach Boys - Pet Sounds (2c2c80be-57e3-44d2-a910-a7a873683141)
Computing track assignment...
...done.
Success. Distance: 0.00
Requesting MusicBrainz release 52e75ffe-a425-41ff-b6e0-2146cc89a06f
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_album_extract
Sending event: albuminfo_received
Candidate: The Beach Boys - Pet Sounds (52e75ffe-a425-41ff-b6e0-2146cc89a06f)
Computing track assignment...
...done.
Success. Distance: 0.00
Requesting MusicBrainz release 7dc0f5eb-6b62-3611-bc4b-ef5430017b18
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_album_extract
Sending event: albuminfo_received
Candidate: The Beach Boys - Pet Sounds (7dc0f5eb-6b62-3611-bc4b-ef5430017b18)
Computing track assignment...
...done.
Success. Distance: 0.00
Requesting MusicBrainz release 2ae701fc-5c5a-4540-b93a-2323920f978c
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_album_extract
Sending event: albuminfo_received
Candidate: The Beach Boys - Pet Sounds (2ae701fc-5c5a-4540-b93a-2323920f978c)
Computing track assignment...
...done.
Success. Distance: 0.02
Requesting MusicBrainz release 9de804cf-8a14-44a0-8e10-da47a24e39c3
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_track_extract
Sending event: mb_album_extract
Sending event: albuminfo_received
Candidate: The Beach Boys - Pet Sounds (9de804cf-8a14-44a0-8e10-da47a24e39c3)
Computing track assignment...
...done.
Success. Distance: 0.00
Evaluating 5 candidates.

Traceback (most recent call last):
  File "/usr/bin/beet", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/usr/lib/python3.11/site-packages/beets/ui/__init__.py", line 1782, in main
    _raw_main(args)
  File "/usr/lib/python3.11/site-packages/beets/ui/__init__.py", line 1769, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/usr/lib/python3.11/site-packages/beets/ui/commands.py", line 1285, in import_func
    import_files(lib, paths, query)
  File "/usr/lib/python3.11/site-packages/beets/ui/commands.py", line 1225, in import_files
    session.run()
  File "/usr/lib/python3.11/site-packages/beets/importer.py", line 353, in run
    pl.run_parallel(QUEUE_SIZE)
  File "/usr/lib/python3.11/site-packages/beets/util/pipeline.py", line 446, in run_parallel
    raise exc_info[1].with_traceback(exc_info[2])
  File "/usr/lib/python3.11/site-packages/beets/util/pipeline.py", line 311, in run
    out = self.coro.send(msg)
          ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/beets/util/pipeline.py", line 170, in coro
    task = func(*(args + (task,)))
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/beets/importer.py", line 1469, in user_query
    task.choose_match(session)
  File "/usr/lib/python3.11/site-packages/beets/importer.py", line 905, in choose_match
    choice = session.choose_match(self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/beets/ui/commands.py", line 984, in choose_match
    path_str = ui.colorize('import_path', path_str0)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/beets/ui/__init__.py", line 608, in colorize
    raise ValueError("no such ANSI code %s", code)
ValueError: ('no such ANSI code %s', 'g')
```

```
library: /blk/black/mysys/beet/beets_library.db
directory: /blk/black/og

plugins: ['filetote', 'dirfields']
pluginpath: []
filetote:
    # Or copy all non-music files (it does this by default):
    extensions: .*
    # It can also print what got left:
    print_ignored: yes


import:
    write: yes
    copy: no
    move: yes
    link: no
    hardlink: no
    reflink: no
    delete: no
    resume: ask
    incremental: no
    incremental_skip_later: no
    from_scratch: no
    quiet_fallback: skip
    none_rec_action: ask
    timid: yes
    log:
    autotag: yes
    quiet: no
    singletons: no
    default_action: apply
    languages: []
    detail: no
    flat: no
    group_albums: no
    pretend: false
    search_ids: []
    duplicate_keys:
        album: albumartist album
        item: artist title
    duplicate_action: ask
    bell: no
    set_fields: {}
    ignored_alias_types: []

clutter: ["Thumbs.DB", ".DS_Store"]
ignore: [".*", "*~", "System Volume Information", "lost+found"]
ignore_hidden: yes

replace:
    # Replace bad characters with -
    # prohibited in many filesystem paths
    # |
    '\|': "-"
    # prohibited in many filesystem paths
    # Replace : with ` -`
    ':': " -"
    #  path separators: \ or /
    '[\\/]': "-"
    # Replace double quotation mark with '' (2 apostrophes)
    '\"': "''"
    # Replace bad characters with _
    # prohibited in many filesystem paths
    '[<>\?\*]': _
    # starting and closing periods
    '^\.': _
    '\.$': _
    # control characters
    '[\x00-\x1f]': _
    # dash at the start of a filename (causes command line ambiguity)
    '^-': _
    # Replace bad characters with nothing
    # starting and closing whitespace
    '\s+$': ''
    '^\s+': ''
path_sep_replace: _
drive_sep_replace: _
asciify_paths: true
art_filename: front
max_filename_length: 255

aunique:
    keys: albumartist album
    disambiguators: albumtype year label catalognum albumdisambig releasegroupdisambig
    bracket: '[]'

sunique:
    keys: artist title
    disambiguators: year trackdisambig
    bracket: '[]'

overwrite_null:
  album: []
  track: []

threaded: yes
timeout: 5.0
per_disc_numbering: yes
verbose: 0
terminal_encoding:
original_date: no
artist_credit: no
id3v23: no
va_name: "Various Artists"

ui:
    terminal_width: 80
    length_diff_thresh: 10.0
    color: yes
    colors:
        text_success: green
        text_warning: yellow
        text_error: red
        text_highlight: red
        text_highlight_minor: lightgray
        action_default: turquoise
        action: blue

format_item: $artist - $album - $title
format_album: $albumartist - $album
time_format: '%Y-%m-%d %H:%M:%S'
format_raw_length: no

sort_album: albumartist+ album+
sort_item: artist+ album+ disc+ track+
sort_case_insensitive: yes

paths:
    default: $albumartist [$year] - $album%aunique{}/$disc. $disctitle - $track. $title
    # singleton: Non-Album/$artist/$title
    # comp: Compilations/$album%aunique{}/$track $title

statefile: state.pickle

musicbrainz:
    enabled: yes
    host: musicbrainz.org
    https: no
    ratelimit: 1
    ratelimit_interval: 1.0
    searchlimit: 5
    extra_tags: []
    genres: no

match:
    strong_rec_thresh: 0.04
    medium_rec_thresh: 0.25
    rec_gap_thresh: 0.25
    max_rec:
        missing_tracks: medium
        unmatched_tracks: medium
    distance_weights:
        source: 2.0
        artist: 3.0
        album: 3.0
        media: 1.0
        mediums: 1.0
        year: 1.0
        country: 0.5
        label: 0.5
        catalognum: 0.5
        albumdisambig: 0.5
        album_id: 5.0
        tracks: 2.0
        missing_tracks: 0.9
        unmatched_tracks: 0.6
        track_title: 3.0
        track_artist: 2.0
        track_index: 1.0
        track_length: 2.0
        track_id: 5.0
    preferred:
        countries: []
        media: []
        original_year: no
    ignored: []
    required: []
    ignored_media: []
    ignore_data_tracks: yes
    ignore_video_tracks: yes
    track_length_grace: 10
    track_length_max: 30
```

JOJ0 commented:
Awesom investigation @RollingStar, I agree that some error handling for legacy config handling is desired.

I would also hve the idea that the user is warned that they should either remove or update their legacy config.

RollingStar commented:
@JOJ0 
> I would also hve the idea that the user is warned that they should either remove or update their legacy config.

Would be nice in other cases as well. For example the convert plugin has a weird negative config that should be inverted (never_convert_lossy_files -> convert_lossy)

RollingStar commented:
Possible bug in mixed-medium releases (ex. 1 CD and 1 vinyl record)

https://musicbrainz.org/release/581db767-5221-424b-b6a0-d5db2ff707a1

![image](https://github.com/beetbox/beets/assets/4810902/35baab09-8848-4753-bc64-4482ebda7bf4)

RollingStar commented on beets/ui/commands.py:
```diff
@@ -255,32 +262,180 @@ def penalty_string(distance, limit=None):
+        media = self.match.info.media or "Media"
+        # Build output string.
+        if self.match.info.mediums > 1 and track_info.disctitle:
+            return f"* {media} {track_info.medium}: {track_info.disctitle}"
```
Possible bug in this function when multiple mediums are in one release. See #3721

JOJ0 commented:
> Awesom investigation @RollingStar, I agree that some error handling for legacy config handling is desired.
> 
> I would also hve the idea that the user is warned that they should either remove or update their legacy config.

I fixed that issue in the commit I just pushed. FYI @Serene-Arc since we talked about what's left to do. So that is done! :-)

@RollingStar the new UI code already has a legacy config check. It actually does not error out when it finds legacy config but translates it to the new config format (where each defined color must be a yaml list). The check for "is it string or list" was just outdated (Python2)

Ok so that part is out of the way. IMHO the only must have for a merge is at least some rudimentary documentation about the new options.

RollingStar commented on docs/reference/config.rst:
```diff
@@ -427,20 +427,66 @@ the ``color`` option is set to ``yes``. For example, you might have a section
+length_diff_thresh
+~~~~~~~~~~~~~~~~~~
+
+FIXME::
```
Looks like it's maximum length in seconds that the importer UI does not consider important. Above this limit, the importer will use text_highlight to note the differences in track runtime between the proposed metadata and the existing track. It does not appear to affect the autotagger matching or difference score calculation in matching at all. https://github.com/beetbox/beets/blob/e10b955a931e4c205b0cadf0860797c0aeee736c/beets/ui/commands.py#L391

RollingStar replied:
Can we change the code to >= ? it feels more intuitive.

RollingStar replied:
For the docs, something like "Beets compares the length of the imported track with the length stored in the proposed metadata provider (Musicbrainz, Discogs, etc.). If any tracks differ by at least `length_diff_thresh` seconds, they will be colored with `text_highlight`. Below this threshold, different track lengths are colored with text_highlight_minor."

JOJ0 replied:
Thanks so much for investigating and describing. That is indeed very helpful! Awesome teamwork here! 
I wanted to shorten the paragraph more but failed. It's almost exactly what you suggested:
![Screenshot 2023-10-12 at 18 35 05](https://github.com/beetbox/beets/assets/2733783/6297d88b-ac27-4290-b437-ef99a5884b41)



You agree were I link to in the very last braces (see ....)
Unfortunately `distance_weights` is undocumented. I'm not sure how to properly link to config_default.yaml - or if that should better be fixed up nicely by including that config directive properly in the docs.....

RollingStar replied:
Distance_weights is undocumented, I say just mark it in `monospace` and leave the documentation to another day.

RollingStar replied:
We would need to change the code from > to >= when comparing length_diff_thresh to match the doc (and what is more intuitive imo). https://github.com/beetbox/beets/blob/3024e6eb425d37e76f0b3ddc93369aea38ad93ad/beets/ui/commands.py#L496

Regarding doc length, yes I am verbose, but I've never wished a doc was shorter as a user. :)

JOJ0 replied:
@Serene-Arc should we change that as @RollingStar suggested?

JOJ0 replied:
Both of you FYI @RollingStar @Serene-Arc I fixed that as suggested. It now fits with the docs and I agree it is now more intuitive.

I was curious now @RollingStar, since I was not sure how the word "threshold" is actually defined. This dictionary description implies a "greather than" though:

"the level or point at which you start to experience something, or at which something starts to happen"

https://dictionary.cambridge.org/us/dictionary/english/threshold#google_vignette

Serene-Arc replied:
I don't think there's any harm in changing that. I agree that it makes more intuitive sense if it's `>=` and it's so minor that most people won't even notice. As long as all of the tests pass, it should be fine.

Serene-Arc commented:
I added a short section on the remaining FIXME in the docs. It should be rather self-explanatory, if I've understood it correctly.

JOJ0 commented:
Hmm, I was wondering what column and newline really is? What changes exactly. I dont remember

Serene-Arc commented:
Those are the two options given in the code. They call different functions which are in `ui/__init__.py`.

JOJ0 commented:
Sorry I just force-pushed to correct my crap commit wording. I'm sorry. Was not a good idea since we are working both on the branch right now!!!

JOJ0 commented:
> Those are the two options given in the code. They call different functions which are in `ui/__init__.py`. I can add more detail in the section if you want but they're layouts so the best thing might be for the users to try them both and see which one they prefer. A visual will be much better than whatever description we can provide, unless they have very different intended use cases.

Yeah let's leave the layout thing as you briefly mentioned it. It's not really explained but just mentioning that it can be changed is enough to encourage the user to play with it. I'm fine with that. Good enough! Thanks!

What I'm not fine with is the second sentence of the indentations - I think there is something gramatically wrong - or the sentence should be 2 sentences. I'll try to rephrase. Give me a minute.

But then.....I think....I'll hit that button....... .-))))) Thanks for pushing forward here @Serene-Arc, appreciated! :-)))

Serene-Arc commented:
Wonderful! Congrats on really pushing to get this done, you got it to the finish line!

Serene-Arc commented:
@JOJ0 The `::` needs to be on a new paragraph in the section as I wrote, as it was. Otherwise it will add a colon to the end of the sentence, making it `.:`.

JOJ0 commented:
@Serene-Arc after resolving the conflict because of the just-merged #4547 I'm getting this error in the new UI: You know by heart if I failed with the conflict-resolving or if it really is an incompatibility with the new UI?

```
    Match (90.0%):
    Fat Freddy's Drop - LOCK-IN
    ≠ year, tracks
Disambiguation string key albumrelease does not exist.
    Spotify, None, 2020, None, THE DROP, None, None
    https://open.spotify.com/album/6D8fIhNTVPMe7DeOcJ4gfg
    * Artist: Fat Freddy's Drop
    * Album: LOCK-IN
       ≠ (#1) Special Edition (LOCK-IN) - F - 90 (7:31) -> (#1) Special Edition - LOCK-IN (7:31)
```

Serene-Arc commented:
Almost certainly the latter. I wrote that PR so I'll take a look.

Serene-Arc commented:
Might have to wait a few hours, Musicbrainz doesn't seem to be responding correctly.

Serene-Arc commented:
Got it working, I can't for the life of me remember what that option was supposed to be but it doesn't exist on the album info object, so I removed it from the default configuration. In other words, all good now.

JOJ0 commented:
> Got it working, I can't for the life of me remember what that option was supposed to be but it doesn't exist on the album info object, so I removed it from the default configuration. In other words, all good now.

Awesome, thanks for fixing so quickly! OMG, I'll hit that b....

JOJ0 commented:
OMG, 8 (in words: "Eight") years later....beets finally has an updated importer UI.

Thanks to @RollingStar, @Serene-Arc for the awesome teamwork! @sampsyo and @wisp3rwind,  @arsaboo and @wazabees for testing, all the others I'm sure I forgot to mention that tested or sent feedback and everyone else I forgot!

But most of all, thanks to @mxmerz for the initial implementation, and @davidswarbrick for bringing it back to life and fixing some rather nasty, tricky bugs!!!

:-3 :-3 :-3

arsaboo commented:
This is awesome guys....thanks to everyone who contributed to (and more importantly persisted with) this. Let us keep an eye out for bugs as more folks test the new UI.

mxmerz commented:
Haha, so apparently it took eight (or more) people eight years to clean up my messy code 😉 But in all seriousness – thanks to you all! Truly incredible, I can't wait to try it out!

davidswarbrick commented:
Thanks to all who contributed, reviewed, documented, reported bugs and fixed (& continues to fix) them! So excited to see these changes in the main branch and get them out into the wider beets community!!

catap commented:
@davidswarbrick I think I found a regression in this changes. When someoen uses non dark theme, like me for last couple of weeks, it leads to almost not used UI. See:

![image](https://github.com/beetbox/beets/assets/37775/5b680102-ecdb-41ad-b8fb-4716b0bc9573)

You may repeat it by used Xterm with this theme: https://github.com/janoamaral/Xresources-themes/blob/master/light/PaperColor.Xresources

Serene-Arc commented:
Hi @catap, if you could open a bug for this, it'll be easier to track.

catap commented:
@Serene-Arc here it is: https://github.com/beetbox/beets/issues/5342

arsaboo commented:
Colors can be configured: https://beets.readthedocs.io/en/stable/reference/config.html#colors
